src/main/kotlin/com/cultureamp/eventsourcing/BlockingAsyncEventProcessorWaiter.kt

```diff
@@ -0,0 +1,47 @@
+package com.cultureamp.eventsourcing
+
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withTimeout
+
+class BlockingAsyncEventProcessorWaiter<M : EventMetadata>(
+    private val eventProcessors: List<BookmarkedEventProcessor<M>>,
+    private val maxWaitMs: Long,
+    private val pollWaitMs: Long,
+    private val logger: (String) -> Unit = System.out::println,
+) {
+    fun waitUntilProcessed(events: List<SequencedEvent<M>>) {
+        val projectorToMaxRelevantSequence: Map<BookmarkedEventProcessor<M>, Long> = eventProcessors.associateWith { eventProcessor ->
+            val eventProcessorEventTypes = eventProcessor.sequencedEventProcessor.domainEventClasses().toSet()
+            val relevantSequencedEvents = events.filter { eventProcessorEventTypes.contains(it.event.domainEvent::class) }
+            val maxRelevantSequence = relevantSequencedEvents.lastOrNull()?.let { it.sequence }
+            maxRelevantSequence
+        }.filterNotNullValues()
+        if (projectorToMaxRelevantSequence.isNotEmpty()) {
+            runBlocking {
+                withTimeout(maxWaitMs) {
+                    while (anyLaggingFrom(projectorToMaxRelevantSequence)) {
+                        delay(pollWaitMs)
+                    }
+                }
+            }
+        }
+    }
+
+    private fun anyLaggingFrom(projectorToMaxRelevantSequence: Map<BookmarkedEventProcessor<M>, Long>): Boolean {
+        val bookmarkStoreToBookmarkNames = projectorToMaxRelevantSequence.keys.map { it.bookmarkStore to it.bookmarkName }.groupBy { it.first }.mapValues { it.value.map { it.second }.toSet() }
+        val bookmarks = bookmarkStoreToBookmarkNames.flatMap { it.key.bookmarksFor(it.value) }.toSet()
+        val desired = projectorToMaxRelevantSequence.mapKeys { it.key.bookmarkName }
+        val actual = bookmarks.associate { it.name to it.sequence }
+        val diff = desired.map { it.key to it.value - (actual[it.key] ?: 0) }.toMap()
+        val lagging = diff.filter { it.value > 0 }
```

Comment:
same here: this code is pretty dense and hard to read. no specific suggestion though

Reply:
I'm hoping the variable names at least help document what's going on - I got used to reading code like this in my Scala days so it doesn't hurt my brain as badly and I just consider it functional data transform smoosh blahblah